### PR TITLE
Fix writing dictionary data to parquet

### PR DESCRIPTION
**Pull request by czentgr**
In a chunked approach the partitioned data is sliced into
dictionary vectors where the partition vectors are shared
across chunks. A previous chunk row position is changed and
not applicable to the prior chunk when the dictionary content
is resolved. This results in incorrect data being written
to the partitioned files.

In addition, Arrow does not support NULL values in a
dictionary and throws an exception.
The exception is
NotImplemented: Writing DictionaryArray with null encoded in dictionary type not yet supported
See https://github.com/apache/arrow/blob/73589ddd60e4cbcd860102871692541989ea38c6/cpp/src/parquet/arrow/path_internal.cc#L752

To solve both issues, the dictionary vector representing
the partitioning is flattened into a FlatVector. 
As a result the data is copied to persist it
across chunks.

To make the constant vectors in the TableWriterTest work
they are also flattened into flat vectors with the same
logic the dictionary vectors are.

The ArrowBridge has a new optional option structure to
indicate if dictionary and constant vectors should be flattened.

Resolves https://github.com/facebookincubator/velox/issues/5560

### CODE REVIEW

File: velox/vector/arrow/Bridge.cpp [hunk 1]
```diff
@@ -101,6 +107,9 @@ class VeloxToArrowBridgeHolder {
   std::unique_ptr<ArrowArray*[]> children_;
 
   std::unique_ptr<ArrowArray> dictionary_;
+
+  // Keep track of used base vectors for dictionary processing.
+  std::vector<std::shared_ptr<velox::BaseVector>> baseVectorPool_;
```
Comment (czentgr) [inline velox/vector/arrow/Bridge.cpp:112]: Replace `std::shared_ptr<velox::BaseVector>` with `VectorPtr`.

Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:112]: Change naming to `flatVectorPool_` and update comment.

File: velox/exec/tests/TableWriteTest.cpp [hunk 2]
```diff
@@ -1006,9 +1006,14 @@ class PartitionedTableWriterTest
...
     std::vector<uint64_t> testParams;
-    const std::vector<bool> multiDriverOptions = {false, true};
-    // Add Parquet with https://github.com/facebookincubator/velox/issues/5560
+    std::vector<bool> multiDriverOptions = {false, true};
     std::vector<FileFormat> fileFormats = {FileFormat::DWRF};
+    if (hasWriterFactory(FileFormat::PARQUET)) {
+      fileFormats.push_back(FileFormat::PARQUET);
+      // Add Parquet with https://github.com/facebookincubator/velox/issues/5560
+      // Temporarily remove multiDriver == true from the Parquet run.
+      multiDriverOptions.pop_back();
```
Comment (majetideepak) [inline velox/exec/tests/TableWriteTest.cpp:1015]: Won't this remove the multiDriver option for DWRF as well?
We should simply add Parquet here and skip below inside the loop when 
`multiDriver = true` and `fileFormat = Parquet`.
Can you also update the [issue](https://github.com/facebookincubator/velox/issues/5560) with the error when the above condition holds?

File: velox/exec/tests/TableWriteTest.cpp [hunk 3]
```diff
@@ -1118,9 +1123,14 @@ class BucketedTableOnlyWriteTest
...
```
Comment (majetideepak) [inline velox/exec/tests/TableWriteTest.cpp:1132]: same as above.

File: velox/vector/arrow/Bridge.cpp [hunk 4]
```diff
@@ -559,24 +568,39 @@ void exportMaps(
...
+  auto dictVector = vec.asUnchecked<DictionaryVector<NativeType>>();
+  rows.apply([&](vector_size_t row) {
+    if (dictVector->isNullAt(row)) {
+      flatVector->setNull(row, true);
+    } else {
+      flatVector->set(row, dictVector->valueAt(row));
+    }
+  });
+  holder.saveVector(flatVector);
+  exportBase(*flatVector, rows, out, pool);
```
Comment (majetideepak) [inline velox/vector/arrow/Bridge.cpp:590]: can we `exportFlat` directly here?

File: velox/vector/arrow/Bridge.cpp [hunk 5]
```diff
@@ -101,6 +107,9 @@ class VeloxToArrowBridgeHolder {
   std::unique_ptr<ArrowArray*[]> children_;
 
   std::unique_ptr<ArrowArray> dictionary_;
+
+  // Keep track of used base vectors for dictionary processing.
```
Comment (majetideepak) [inline velox/vector/arrow/Bridge.cpp:111]: Can you explain why we cannot use `buffers_` above? Why do we need this additional vector?

**Review by majetideepak — COMMENTED** [review]
@czentgr nice find and fix! Some comments.

File: velox/vector/arrow/Bridge.cpp (hunk 5 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:111]: The idea was that we create a new vector with new memory and copy values.
When the vector goes out pf scope the BufferPtr of the holder points to the values buffer of the vector.
This isn't a problem with the passed in vector that comes from the higher layers.

I assumed I need to keep the vector around so the values are still legit by the time the BufferPtr is written to the file which is as long as the holder is not deleted.
But looks like they are allocated as shared_ptr and then moved to be a member. So looks like we might not need to hold on to the vector as one holder of the BufferPtr for values.

File: velox/vector/arrow/Bridge.cpp (hunk 4 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:590]: No, we need the nulls exported as well. `exportFlat` only exports the values and not the nulls.

File: velox/exec/tests/TableWriteTest.cpp (hunk 2 above)
Comment (czentgr) [reply velox/exec/tests/TableWriteTest.cpp:1015]: Yeah, I tried to be smarter and failed.

**Comment by czentgr** [comment]
Need to address the `velox_arrow_bridge_test` failure now that the dictionary is working differently.

File: velox/exec/tests/TableWriteTest.cpp [hunk 6]
```diff
@@ -1006,11 +1006,18 @@ class PartitionedTableWriterTest
 
   static std::vector<uint64_t> getTestParams() {
     std::vector<uint64_t> testParams;
-    const std::vector<bool> multiDriverOptions = {false, true};
+    std::vector<bool> multiDriverOptions = {false, true};
```
Comment (czentgr) [inline velox/exec/tests/TableWriteTest.cpp:1009]: Undo this change.

File: velox/exec/tests/TableWriteTest.cpp [hunk 7]
```diff
@@ -1118,11 +1125,18 @@ class BucketedTableOnlyWriteTest
...
```
Comment (czentgr) [inline velox/exec/tests/TableWriteTest.cpp:1128]: Undo this change.

File: velox/vector/arrow/Bridge.cpp (hunk 4 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp]: Added an an optimized path for non-nulls.

**Comment by czentgr** [comment]
Code is being reworked to remove the change from the bridge into the writer.

File: velox/exec/tests/TableWriteTest.cpp [hunk 8]
```diff
@@ -137,7 +138,7 @@ struct TestParam {
 
   CompressionKind compressionKind() const {
     return static_cast<facebook::velox::common::CompressionKind>(
-        (value & ((1L << 48) - 1)) >> 40);
+        (value & ((1L << 56) - 1)) >> 48);
```
Comment (czentgr) [inline velox/exec/tests/TableWriteTest.cpp:141]: To be removed later because this is addressed in a different PR which is out for review.

File: velox/exec/tests/TableWriteTest.cpp [hunk 9]
```diff
@@ -2715,6 +2726,10 @@ TEST_P(AllTableWriterTest, columnStatsWithTableWriteMerge) {
 // TODO: add partitioned table write update mode tests and more failure tests.
 
 TEST_P(AllTableWriterTest, tableWriterStats) {
+  if (fileFormat_ == FileFormat::PARQUET) {
```
Comment (czentgr) [inline velox/exec/tests/TableWriteTest.cpp:2729]: To be removed depending on another PR fixing the stats collection.

File: velox/exec/tests/TableWriteTest.cpp (hunk 8 above)
Comment (majetideepak) [reply velox/exec/tests/TableWriteTest.cpp:141]: That PR got merged. Let's remove it.

**Review by majetideepak — COMMENTED** [review]
@czentgr is this PR ready?

**Comment by majetideepak** [comment]
Please update the comment with the unsupported Arrow Dictionary with nulls issue as well. That is the major reason to support this change. A link to the Arrow code where the error is thrown will be helpful.

File: velox/vector/arrow/Bridge.cpp [hunk 10]
```diff
@@ -576,14 +614,53 @@ void exportDictionary(
...
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool,
+    VeloxToArrowBridgeHolder& holder) {
+  VELOX_CHECK(vec.isScalar(), "Flattening is only supported for scalar types.");
+  VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+      flattenAndExport, vec.typeKind(), vec, rows, out, pool, holder);
+}
+
+void exportDictionary(
```
Comment (majetideepak) [inline velox/vector/arrow/Bridge.cpp:632]: Separate functions for `exportDictionary` and `exportConstant` can be avoided. Let's inline the body inside `exportBase`.

**Review by majetideepak — COMMENTED** [review]
Remaining changes look good

File: velox/vector/arrow/Bridge.cpp (hunk 10 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:632]: Yes, I had considered this but though it was better to have a wrapper. I'll change it.

**Comment by czentgr** [comment]
@majetideepak latest run looks clean. I applied all review comments so far.

File: velox/vector/arrow/Bridge.cpp [hunk 11]
```diff
@@ -554,11 +562,41 @@ void exportMaps(
...
+      } else {
+        flatVector->set(row, decoded.valueAt<NativeType>(row));
+      }
+    });
+    exportBase(*flatVector, rows, out, pool, {false, false});
+  } else {
+    allRows.applyToSelected([&](vector_size_t row) {
+      flatVector->set(row, decoded.valueAt<NativeType>(row));
+    });
+    exportFlat(*flatVector, rows, out, pool, holder);
```
Comment (majetideepak) [inline velox/vector/arrow/Bridge.cpp:596]: Since we specialize here for each type, why not use `exportValues/exportStrings` directly?
Or I wonder if there is another way to flatten a dictionary without having to specialize for each type.

File: velox/vector/arrow/Bridge.cpp [hunk 12]
```diff
@@ -554,11 +562,41 @@ void exportMaps(
...
+
+  if (decoded.mayHaveNulls()) {
+    allRows.applyToSelected([&](vector_size_t row) {
+      if (decoded.isNullAt(row)) {
+        flatVector->setNull(row, true);
+      } else {
+        flatVector->set(row, decoded.valueAt<NativeType>(row));
+      }
+    });
+    exportBase(*flatVector, rows, out, pool, {false, false});
```
[the remark this answers is not in the document]
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:591]: The TableWriteTest has nulls in the data for sure. Hence how we found out about the Dictionary null issue in Arrow.
There are tests that write all supported flat types except HUGEINT (as that is not externalized I suppose).

So I think there is a decent coverage in terms of types and values. I guess one thing could be to try and fuzz the values.

[the remark this answers is not in the document]
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:591]: Maybe we don't need the nulls - they are already exported at the beginning when `exportBase` is called for the column. If we flatten everything then `exportNulls` just redoes the work when calling `exportBase`. So I think I was wrong about the nulls needed to be written when you asked earlier.

File: velox/vector/arrow/Bridge.cpp (hunk 11 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:596]: I mean `exportFlat` is a nice wrapper around the required switch to decide if we call  `exportValues` or `exportStrings`. Now I suppose we could add a big constexpr if (instead of the switch) since the type is coming from the template so it knows which function to call when the template is instantiated.

File: velox/vector/arrow/Bridge.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:591]: Actually, the nulls in the initial exportNulls handle null column values. In the dictionary case, we need to deal with the nulls of the dictionary values after decoding so we still need to process them but could potentially optimize and not call exportBase.

File: velox/vector/arrow/Bridge.cpp (hunk 11 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:596]: The main thing about the flattening is to create a type specific flat vector that holds the values as referenced by in the dictionary - and to be able to read the type specific data from the decoded vector. If there is another option (without bloat - I mean it could be written without a template but ...) I don't see it.

**Review by majetideepak — APPROVED** [review]

File: velox/vector/arrow/Bridge.cpp [hunk 13]
```diff
@@ -790,7 +841,7 @@ void exportConstantValue(
         wrapInBufferViewAsViewer(vec.valuesAsVoid(), bufferSize),
         vec.mayHaveNulls() ? 1 : 0);
   }
-  exportToArrowImpl(*valuesVector, selection, out, pool);
+  exportToArrowImpl(*valuesVector, selection, out, pool, ArrowOptions());
```
Comment (Yuhta) [inline velox/vector/arrow/Bridge.cpp:844]: What if the input is a dictionary inside constant and we want to flatten the inner dictionary but not the outer constant?

File: velox/vector/arrow/Bridge.cpp [hunk 14]
```diff
@@ -721,11 +729,42 @@ void exportMaps(
...
+void flattenAndExport(
+    const BaseVector& vec,
+    const Selection& rows,
+    ArrowArray& out,
+    memory::MemoryPool* pool,
+    VeloxToArrowBridgeHolder& holder) {
+  using NativeType = typename velox::TypeTraits<kind>::NativeType;
+  SelectivityVector allRows(vec.size());
+  DecodedVector decoded(vec, allRows);
+  auto flatVector = BaseVector::create<FlatVector<NativeType>>(
```
Comment (Yuhta) [inline velox/vector/arrow/Bridge.cpp:747]: Why not just call `BaseVector::flattenVector`?

Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:747]: That function does in-place flattening of the input. The input is a const reference so we would have to  modify the input type. 
From a data perspective the input is a dictionary vector that wraps around the actual partitioned data. Doing an in-place change would likely affect the input data (I have not tried to change the code to try and see what would happen).

File: velox/vector/arrow/Bridge.cpp (hunk 13 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:844]: Why would a dictionary encoding be inside a constant encoding? As in, a single dictionary index and value that represents the constant? Would the parquet writer get such a data input?
This change assumes that either encoding wraps around flat data and not another encoding. 
It doesn't currently support this particular case.

The other unsupported case is when the dictionary/constant wraps a complex type.

Comment (Yuhta) [reply velox/vector/arrow/Bridge.cpp:844]: Can you add a check for these cases?

File: velox/vector/arrow/Bridge.cpp (hunk 14 above)
Comment (Yuhta) [reply velox/vector/arrow/Bridge.cpp:747]: It only does the in-place change when the reference is unique.  Otherwise a copy is created same as here.  But yeah with the current function signature it would be hard so we can keep it as it is for now.

File: velox/vector/arrow/Bridge.cpp (hunk 13 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:844]: The check for the scalar types is on line 795. If the wrapped type is not flat then return an error. Is that sufficient?

Comment (Yuhta) [reply velox/vector/arrow/Bridge.cpp:844]: You can add a check that `valueVector` is either null or flat

**Comment by czentgr** [comment]
The CI pipeline on Linux fails the test on the newly added check for the values() vector. This did not occur on Mac. Need to see if I can repro on Linux.

File: velox/vector/arrow/Bridge.cpp (hunk 13 above)
Comment (czentgr) [reply velox/vector/arrow/Bridge.cpp:847]: Adding that constant vector cannot wrap another constant or dictionary at this time. There is a check that prohibits this scenario in the ConstantVector constructor

**Comment by czentgr** [comment]
@Yuhta Can you please provide details on the linter warnings so I can fix them?
